code/modules/mob/living/carbon/life.dm

```diff
@@ -37,13 +37,17 @@
 			else
 				hud_used.healths.icon_state = "health6"
 
-///gives humans oxy when dragged by a xeno, called on COMSIG_MOVABLE_PULL_MOVED
-/mob/living/carbon/human/proc/oncritdrag()
+///gives humans oxy when moved around in certain conditions. called on COMSIG_MOVABLE_MOVED
+/mob/living/carbon/human/proc/on_crit_moved(datum/source, atom/old_loc, movement_dir, forced = FALSE, list/old_locs)
 	SIGNAL_HANDLER
-	if(isxeno(pulledby))
-		if(adjustOxyLoss(HUMAN_CRITDRAG_OXYLOSS)) //take oxy damage per tile dragged
+	if(pulledby || throwing) // only catch the scenarios we're interested in: pulls and throws
+		if(pulledby && !isxeno(pulledby)) // only care about xenos pulling us
 			return
-		INVOKE_ASYNC(src, PROC_REF(adjustBruteLoss), HUMAN_CRITDRAG_OXYLOSS)
+		if(throwing && !isxeno(thrower)) // same here, albeit for throwing
+			return
+		if(!adjustOxyLoss(HUMAN_CRITDRAG_OXYLOSS)) // take oxy damage per tile moved
+			INVOKE_ASYNC(src, PROC_REF(adjustBruteLoss), HUMAN_CRITDRAG_OXYLOSS) // if we can't take oxy damage (for some reason), take it as brute instead
+		updatehealth() // force a health update so we can't get dragged any further than we should be
```

Comment:
the real question is why adjustoxloss doesnt have a updatehealth arg

Reply:
![image](https://github.com/user-attachments/assets/1546e0c0-5af9-4e11-a3d7-5835a895ddb0)